src/runtime/relax_vm/lm_support.cc

```diff
@@ -510,6 +510,68 @@ int SampleTopPFromProb(NDArray prob, double top_p, double uniform_sample) {
 
 TVM_REGISTER_GLOBAL("vm.builtin.sample_top_p_from_prob").set_body_typed(SampleTopPFromProb);
 
+// NOLINTNEXTLINE(runtime/references)
+void LogSoftmax(NDArray logits, const NDArray& output) {
+  /* log_softmax from logits is calculated.
+   * Both operations are joined to more quick and stable calculations:
+   * Log(Softmax(logits))[i] = Log(exp(logits[i])/Sum(exp(logits[i]), i)) =
+   * logits[i] - Log(Sum(exp(logits[i]), i)) =
+   * logits[i] - logits_max - Log(Sum(exp(logits[i]-logits_max), i))
+   */
+  // TODO(vvchernov): torch.log_softmax calculates sum by chunks in parallel and reduces it after,
+  // possibly we need to do it here and extend it to other "vm.builtin.sample_top_p_from..."
+  ICHECK(logits.IsContiguous());
+  ICHECK(logits.DataType() == DataType::Float(32));
+
+  if (logits->device.device_type != kDLCPU) {
+    logits = logits.CopyTo(DLDevice{kDLCPU, 0});
+  }
+
+  ICHECK(logits->device.device_type == kDLCPU);
+
+  // TODO(vvchernov): It is still assumed that batch size = 1 only
+  for (int i = 0; i < logits->ndim - 2; ++i) {
+    ICHECK_EQ(logits->shape[i], 1) << "The leading dimensions of logits must be 1";
+  }
+
+  size_t seq_length = logits->shape[logits->ndim - 2];
+  size_t vocab_length = logits->shape[logits->ndim - 1];
+  std::vector<float> data(vocab_length);
+  float* res_ptr = static_cast<float*>(output->data);
+
+  for (size_t seq_ind = 0; seq_ind < seq_length; ++seq_ind) {
+    data.clear();
```

Comment:
do you needs to clear it? i feel data is not needed. you can just operate on inputs 

and result_log_probs inplace,

First get the max  value, then do the subtraction

Reply:
aka you can use the output array to store the intermediate result of expf and then run the final subtractions

Reply:
Hello @tqchen! I've fixed it